waiter/src/waiter/core.clj

```diff
@@ -1085,21 +1090,26 @@
                              (scheduler/validate-service scheduler service-id)
                              (service/start-new-service
                                scheduler descriptor start-service-cache scheduler-interactions-thread-pool)))
-   :start-work-stealing-balancer-fn (pc/fnk [[:settings [:work-stealing offer-help-interval-ms reserve-timeout-ms]]
+   :start-work-stealing-balancer-fn (pc/fnk [[:settings [:work-stealing offer-help-interval-ms offer-idle-timeout-ms
+                                                         reserve-timeout-ms]]
                                              [:state offers-allowed-semaphore router-id]
-                                             make-inter-router-requests-async-fn router-metrics-helpers]
+                                             enable-work-stealing-support? make-inter-router-requests-async-fn router-metrics-helpers]
                                       (fn start-work-stealing-balancer [populate-maintainer-chan! service-id]
-                                        (let [{:keys [service-id->router-id->metrics]} router-metrics-helpers]
-                                          (work-stealing/start-work-stealing-balancer
-                                            populate-maintainer-chan! reserve-timeout-ms offer-help-interval-ms
-                                            offers-allowed-semaphore service-id->router-id->metrics
-                                            make-inter-router-requests-async-fn router-id service-id))))
+                                        (let [support-work-stealing? (enable-work-stealing-support? service-id)]
+                                          (log/info service-id "has work-stealing support" (if support-work-stealing? "enabled" "disabled"))
+                                          (when support-work-stealing?
+                                            (let [{:keys [service-id->router-id->metrics]} router-metrics-helpers]
+                                              (work-stealing/start-work-stealing-balancer
+                                                populate-maintainer-chan! reserve-timeout-ms offer-help-interval-ms
+                                                offer-idle-timeout-ms offers-allowed-semaphore service-id->router-id->metrics
+                                                make-inter-router-requests-async-fn router-id service-id))))))
    :stop-work-stealing-balancer-fn (pc/fnk []
-                                     (fn stop-work-stealing-balancer [service-id work-stealing-chan-map]
-                                       (log/info "stopping work-stealing balancer for" service-id)
-                                       (async/go
-                                         (when-let [exit-chan (get work-stealing-chan-map [:exit-chan])]
```

Comment:
This is a bug where work-stealing go-blocks are not being terminated correctly.
This may explain high CPU usage over time.

Reply:
Nice find